macros/edr/data_monitoring/monitors_query/column_monitoring_query.sql

```diff
@@ -1,21 +1,30 @@
-{% macro column_monitoring_query(monitored_table_relation, timestamp_column, min_bucket_start, column_obj, column_monitors, where_expression=none) %}
+{% macro column_monitoring_query(monitored_table_relation, timestamp_column, min_bucket_start, column_obj, column_monitors, where_expression=none, time_bucket=none) %}
 
-    {%- set max_bucket_end = "'"~ elementary.get_run_started_at().strftime("%Y-%m-%d 00:00:00")~"'" %}
-    {%- set full_table_name_str = "'"~ elementary.relation_to_full_name(monitored_table_relation) ~"'" -%}
+    {% set full_table_name_str = elementary.quote(elementary.relation_to_full_name(monitored_table_relation)) %}
 
-    with filtered_monitored_table as (
+    {% set bucket_start_datediff_expr %}
+      floor({{ elementary.datediff(min_bucket_start, elementary.cast_as_timestamp(timestamp_column), time_bucket.period) }} / {{ time_bucket.count }}) * {{ time_bucket.count }}
+    {% endset %}
 
+    with buckets as (
+        select edr_bucket_start, edr_bucket_end from ({{ elementary.complete_buckets_cte(time_bucket) }})
+        {% if min_bucket_start -%}
+          where edr_bucket_start >= {{ elementary.cast_as_timestamp(min_bucket_start) }}
+        {%- endif %}
+    ),
+
+    filtered_monitored_table as (
         select {{ column_obj.quoted }}
             {% if timestamp_column -%}
-             , {{ elementary.time_trunc('day', timestamp_column) }} as edr_bucket
+             , {{ elementary.cast_as_timestamp(elementary.dateadd(time_bucket.period, elementary.cast_as_int(bucket_start_datediff_expr), min_bucket_start)) }} as start_bucket_in_data
             {%- else %}
-            , {{ elementary.null_timestamp() }} as edr_bucket
+            , {{ elementary.null_timestamp() }} as start_bucket_in_data
             {%- endif %}
         from {{ monitored_table_relation }}
         where
         {% if timestamp_column -%}
-            {{ elementary.cast_as_timestamp(timestamp_column) }} >= {{ elementary.cast_as_timestamp(min_bucket_start) }}
-            and {{ elementary.cast_as_timestamp(timestamp_column) }} <= {{ elementary.cast_as_timestamp(max_bucket_end) }}
+            {{ elementary.cast_as_timestamp(timestamp_column) }} >= (select min(edr_bucket_start) from buckets)
+            and {{ elementary.cast_as_timestamp(timestamp_column) }} < (select max(edr_bucket_end) from buckets)
```

Comment:
It's a bit harder to calculate the `max_complete_bucket_end` because in order to do that you'd have to start the bucketing from `min_bucket_start` whereas previously you just did today at 00:00.

Technically it should still be possible to do it with Jinja, like getting the min and max value of [this](https://github.com/elementary-data/dbt-data-reliability/blob/7fceb254d82343fc93901505fe98af158ce1b219/integration_tests/run_e2e_tests.py#L21), but I liked how it's contained within the buckets CTE and there's no need to re-calculate it.

Reply:
OK that makes sense to me, didn't think about it
Didn't realize that the "today at 00:00" logic is not correct anymore